amperfied/integrationpluginamperfied.cpp

```diff
@@ -0,0 +1,441 @@
+            thing->setStateValue(energyControlPowerStateTypeId, false);
+        } else {
+            thing->setStateValue(energyControlPowerStateTypeId, true);
+            thing->setStateValue(energyControlMaxChargingCurrentStateTypeId, connection->chargingCurrent());
+        }
+        thing->setStateMinMaxValues(energyControlMaxChargingCurrentStateTypeId, connection->minChargingCurrent(), connection->maxChargingCurrent());
+        thing->setStateValue(energyControlCurrentPowerStateTypeId, connection->currentPower());
+        thing->setStateValue(energyControlTotalEnergyConsumedStateTypeId, connection->totalEnergy() / 1000.0);
+        thing->setStateValue(energyControlSessionEnergyStateTypeId, connection->sessionEnergy() / 1000.0);
+        switch (connection->chargingState()) {
+        case AmperfiedModbusRtuConnection::ChargingStateUndefined:
+        case AmperfiedModbusRtuConnection::ChargingStateA1:
+        case AmperfiedModbusRtuConnection::ChargingStateA2:
+            thing->setStateValue(energyControlPluggedInStateTypeId, false);
+            break;
+        case AmperfiedModbusRtuConnection::ChargingStateB1:
+        case AmperfiedModbusRtuConnection::ChargingStateB2:
+        case AmperfiedModbusRtuConnection::ChargingStateC1:
+        case AmperfiedModbusRtuConnection::ChargingStateC2:
+            thing->setStateValue(energyControlPluggedInStateTypeId, true);
+            break;
+        case AmperfiedModbusRtuConnection::ChargingStateDerating:
+        case AmperfiedModbusRtuConnection::ChargingStateE:
+        case AmperfiedModbusRtuConnection::ChargingStateError:
+        case AmperfiedModbusRtuConnection::ChargingStateF:
+            qCWarning(dcAmperfied()) << "Unhandled charging state:" << connection->chargingState();
+        }
+
+        int phaseCount = 0;
+        if (connection->currentL1() > 1) {
+            phaseCount++;
+        }
+        if (connection->currentL2() > 1) {
+            phaseCount++;
+        }
+        if (connection->currentL3() > 1) {
+            phaseCount++;
+        }
+        if (phaseCount > 0) {
+            thing->setStateValue(energyControlPhaseCountStateTypeId, phaseCount);
+        }
+        thing->setStateValue(energyControlChargingStateTypeId, phaseCount > 0);
+    });
+
+    connection->update();
+
+}
+
+void IntegrationPluginAmperfied::setupTcpConnection(ThingSetupInfo *info)
+{
+    qCDebug(dcAmperfied()) << "setting up TCP connection";
+    Thing *thing = info->thing();
+    NetworkDeviceMonitor *monitor = m_monitors.value(info->thing());
+    AmperfiedModbusTcpConnection *connection = new AmperfiedModbusTcpConnection(monitor->networkDeviceInfo().address(), 502, 1, info->thing());
+
+    connect(connection, &AmperfiedModbusTcpConnection::reachableChanged, thing, [connection, thing](bool reachable){
+        if (reachable) {
+            connection->initialize();
+        } else {
+            thing->setStateValue(connectHomeConnectedStateTypeId, false);
```

Comment:
We should set the current power to 0 if the device is not connected.

    thing->setStateValue(connectHomeCurrentPowerStateTypeId, 0);